### PR TITLE
feat: subcontracting module

### LINKED ISSUE #29604: Subcontracting flow refactoring
### Problem
Currently, the whole subcontracting flow is not intuitive and is sort of a work-around. It even takes a while to figure out the flow.

### Possible Solution
Have an explicit flow for subcontracting:
image

### Possible Changes

**Master level**

- Mark Service item for subcontracting and then map Service Item with FG (BOM)
- Add default subcontractor for the service item
- Purchase Order (PO)
- Mark PO for subcontracting
- On selecting the service item (show the RMs and the cost of the RMs if the item mapping is done) or manually select a BOM for every service item
- Option to create Subcontracting Order (SCO) from a submitted PO

**Subcontracting Order (SCO)**

- Create from PO
- Create independently
- Select supplier
- Select FG items (if started from PO then populate basis the service item mapping)
- View the RMs and the cost
- Add warehouse info for material transfer
- Status: Draft/Open/Partially Received/Completed/Material transferred/Partial Material Transferred
- Option to create Stock entry (material transfer) on a submitted SCO
- Option to create Subcontracting Receipt

**Material Transfer**

- Create from submitted SCO
- Independently create Stock Entry and provide reference for SCO
- Fetch all the RMs and the quantity
- Provide From and To warehouse information
- Transfer partially or full quantity
- Transfer additional material
- Update status of SCO basis the material transferred.
- Eway bill for transfer?

**Subcontracting Receipt (SCR)**

- Create from SCO
- Create independently and provide a reference for SCO (fetch all information)
- Perform quality inspection (settings could be configured)
- Partially receive -update SCO status
- Complete receipt - update SCO status
- Create return basis the quality inspection - update SCO status
- Stock update

**Create Purchase Receipt**

- From PO or SCO or independently by providing the reference for the PO
- Update PO status.

### PR DESCRIPTION
**Pull request by s-aga-r**
### Problem
Currently, the whole subcontracting flow is not intuitive and is sort of a work-around. It even takes a while to figure out the flow.

### Possible Solution
Have an explicit flow for subcontracting:
image

### For QA Team
<details>

Test site: https://subcontracting.frappe.cloud

#### 1. Through Production Plan:
- Create Items
- Create BOMs
- Create a Production Plan
- Create Work Order and Purchase Order (Create > Work Order / Subcontract PO)
- Follow the Manufacturing cycle (Submit the Work Order)
- Set Service Item, Qty and Rate in Purchase Order and Submit it.
- Create Subcontracting Order against the Purchase Order.
- Transfer the raw materials to the Subcontractor.
- Create Subcontracting Receipt against the Subcontracting Order.
- Create the Purchase Receipt / Invoice (Optional) against the Purchase Order.
- Complete the Manufacturing Cycle.

#### 2. Directly create the Purchase Order:
- Create Purchase Order (Select Service Item, Qty, Rate, Finished Good Item and Finished Good Item Qty).
- Create Subcontracting Order against the Purchase Order.
- Supply raw materials to the supplier.
- Create Subcontracting Receipt.
- Create Purchase Receipt / Invoice (Optional)

_NOTE:_ Test data is already on the test site.
</details>

### Possible Changes

**DocType(s)**

**New**

- [x] Subcontracting Order
- [x] Subcontracting Order Service Item (child)
- [x] Subcontracting Order Item (child)
- [x] Subcontracting Order Supplied Item (child)
- [x] Subcontracting Receipt 
- [x] Subcontracting Receipt Item (child)
- [x] Subcontracting Receipt Supplied Item (child)

**To Be Removed (Readonly for Now)**

- [x] Purchase Order Item Supplied (child)

**Master level**

- [x] ~~Mark Service item for subcontracting and then map Service Item with FG (BOM)~~
- [x] Add default subcontractor for the service item
- [x] ~~Purchase Order (PO)~~
- [x] Mark PO for subcontracting
- [x] ~~On selecting the service item (show the RMs and the cost of the RMs if the item mapping is done) or manually select a BOM for every service item~~
- [x] Option to create Subcontracting Order (SCO) from a submitted PO

**Subcontracting Order (SCO)**

- [x] Create from PO
- [x] Create independently (PO*)
- [x] Select supplier
- [x] ~~Select FG items (if started from PO then populate basis service item mapping)~~
- [x] Add warehouse info for material transfer
- [x] Status: Draft/Open/Partially Received/Completed/Material transferred/Partial Material Transferred
- [x] Option to create a Stock entry (material transfer) on a submitted SCO
- [x] Option to create Subcontracting Receipt

**Material Transfer**

- [x] Create from submitted SCO
- [x] Independently create Stock Entry and provide a reference for SCO
- [x] Fetch all the RMs and the quantity
- [x] Provide From and To warehouse information
- [x] Transfer partially or full quantity
- [x] Transfer additional material
- [x] Update status of SCO basis of the material transferred.
- [x] ~~Eway bill for transfer?~~

**Subcontracting Receipt (SCR)**

- [x] Create from SCO
- [x] Create independently and provide a reference for SCO (fetch all information)
- [x] Perform quality inspection (settings could be configured)
- [x] Partially receive -update SCO status
- [x] Complete receipt - update SCO status
- [x] ~~Create return basis the quality inspection - update SCO status~~
- [x] Stock update

**Create Purchase Receipt**

- [x] ~~From PO or SCO or independently by providing the reference for the PO~~
- [x] Update PO status.

### CODE REVIEW

File: erpnext/buying/doctype/purchase_order/purchase_order.py [hunk 1]
```diff
@@ -193,12 +191,25 @@ def validate_minimum_order_qty(self):
...
+	def validate_fg_item_for_subcontracting(self):
 		if self.is_subcontracted:
 			for item in self.items:
-				if not item.bom:
+				if not item.fg_item:
 					frappe.throw(
-						_("BOM is not specified for subcontracting item {0} at row {1}").format(
+						_("Finished Good Item is not specified for service item {0} at row {1}").format(
+							item.item_code, item.idx
+						)
```
Comment (nabinhait) [inline erpnext/buying/doctype/purchase_order/purchase_order.py]:
```suggestion
						_("Row #{0}: Finished Good Item is not specified for service item {1}").format(
							 item.idx, item.item_code
						)
```

Comment (nabinhait) [reply erpnext/buying/doctype/purchase_order/purchase_order.py]: Use this pattern to mention row in all messages

**Review by nabinhait — CHANGES_REQUESTED** [review]

**Comment by s-aga-r** [comment]
Failing checks are not related #[4061143](https://github.com/frappe/erpnext/pull/30955/commits/406114374bbee4831df5fdaa3a49d8980e703d86)

**Comment by s-aga-r** [comment]
Will fix **TEST 2** #[e4b1d6f](https://github.com/frappe/erpnext/pull/30955/commits/e4b1d6f0cb0f274b1d493bf69586d6029e5ea498)

**Comment by rohitwaghchaure** [comment]
@s-aga-r 
Done with the testing, everything is working fine. Just found few minor issues which are as follows

1) If purchase receipt created against purchase order with new flow then system throws an error "Supplier Warehouse" is mandatory. This field should be mandatory only if the purchase receipt made against the old flow purchase order

2) Created Purchase Order (new flow) for 4 quantity and then created subcontracting order from purchase order. System has pulled items details from purchase order with quantity as 1. Ideally quantity should be 4, even system was not allowing to edit the quantity from 1 to 4.

3) I can able to create n number of Subcontracting Purchase Orders against the main purchase order. The button "Create Subcontracting Order" should be hidden after creating of Subcontracting Purchase Orders

**Comment by s-aga-r** [comment]
> For the 2nd issue

There are two cases:

1. When Service Item Qty and FG Item Qty are different.

https://user-images.githubusercontent.com/63660334/176882414-ab34af36-f3d8-418f-9e43-e1d4b31042b0.mov

2. When Service Item Qty and FG Item Qty are equal. 

https://user-images.githubusercontent.com/63660334/176882431-1421cca9-8daf-404f-97fd-cd3db66d138c.mov

**Comment by rohitwaghchaure** [comment]
@s-aga-r 

While checking valuation I found one issue that the rate is different in Subcontracting PO, PR

Subcontracting PO :- Rate was set as 1000
Screenshot 2022-07-01 at 4 31 23 PM

Stock Transfer :- Rate was 1000
Screenshot 2022-07-01 at 4 32 25 PM

Subcontracting PR :-  Rate set to 640, 
Screenshot 2022-07-01 at 4 32 03 PM

If rate we are calculating as per FIFO then rate should be 400 (2*100 + 1000 = 1200 / 3) and not 640
Before subcontracting PR the Stock Queue was "[[2.0, 100.0], [3.0, 1000.0]]"
Screenshot 2022-07-01 at 4 35 42 PM

**Comment by rohitwaghchaure** [comment]
> > For the 2nd issue
> 
> There are two cases:
> 
> 1. When Service Item Qty and FG Item Qty are different.
> 
>  Screen.Recording.2022-07-01.at.4.29.11.PM.mov 
> 2. When Service Item Qty and FG Item Qty are equal.
> 
>  Screen.Recording.2022-07-01.at.4.30.03.PM.mov

@s-aga-r Ok Understood
can we not set default qty as 1 and let user manually set the "Finished Good Item Qty"? or on change of Qty field can we set same qty to "Finished Good Item Qty" field? if user wants to change then they will change it manually.

**Comment by s-aga-r** [comment]
> > > For the 2nd issue
> > 
> > 
> > There are two cases:
> > 
> > 1. When Service Item Qty and FG Item Qty are different.
> > 
> > Screen.Recording.2022-07-01.at.4.29.11.PM.mov
> > 2. When Service Item Qty and FG Item Qty are equal.
> > Screen.Recording.2022-07-01.at.4.30.03.PM.mov
> 
> @s-aga-r Ok Understood can we not set default qty as 1 and let user manually set the "Finished Good Item Qty"? or on change of Qty field can we set same qty to "Finished Good Item Qty" field? if user wants to change then they will change it manually.

Both Qty and FG Item Qty are by default 1.

https://user-images.githubusercontent.com/63660334/176908324-29d06336-2710-47ff-9a9f-aa1970419667.mov

**Pull request merged**